### DISCUSSION TITLE
saas 常见问题

### DISCUSSION THREAD

**Discussion by Harry-Chen** [physics-data/faq#92] [Q&A] [answered]
> GitHub 上评测报告 Broken pipe

可以无视此报错，不影响评分。

额外的，请检查 Action Log 中你的脚本的输出，注意是否用到了评测机上没有安装的指令。如果有非常重要的指令没办法用别的替代，请在 Discussion 中提问。

> Grader 在结果错误时依旧计分

评测脚本内存在一些错误，请使用 `git apply` 应用[这个 Patch](https://gist.github.com/CircuitCoder/6f0103f98c2c2266f615bf5b9bd4540d) （直接使用贴进标准输入即可）

额外的，你也可以直接从 https://github.com/Physics-Data/tpl_saas 更新 `/scripts` 目录

> curl 报错 `curl: (1) Received HTTP/0.9 when not allowed`

通常是你的 HTTP 响应不符合格式，请仔细阅读文档。尤其不能缺少开头的 `HTTP/1.1`。

> curl 访问我的服务器什么也没答应

这是预期的，可能你并没有返回任何响应正文，只有 header。使用 `curl -v` 或者 `curl -i` 可以查看 header。

> macOS 上 `read -N` 报错

理论上你不会用到这个命令，如果需要读取一定数量的（任意）数据，请使用 `head`, `dd` 等命令，`read` 的行为可能是不正确的。特别地，如果你读取的内容中有任何换行等空白字符，或者不可打印字符（特别是 `\0`），`read` 一定会给出错误的结果。

> `GET /api?xxx=yyy` 如何处理

我们按照精确匹配，请报告 404.

**Comment by mhliu0001** [comment, marked as answer]
分享一些我踩过的坑：
1. 注意变量引用需要`$`符号，变量赋值不需要
2. 使用`echo`的时候小心是否需要`-n`（取消末尾的换行符）与`-e`（转义\n等）
3. 使用`$(blablah)`赋值，会将`blablah`的结果中的换行符转成空格
4. 如果出现一些你无法理解的错误，比如说无法重复的错误，请查看是否运行了多个saas.sh，或者端口是否被占用。有用的命令：`ps -e | grep saas.sh`，`lsof -i:1080`，`kill`
5. HTTP请求中的"\r"会造成很多问题，`xxd`是个好东西
6. HTTP请求最后不会有换行符，因此读到最后一行时`read -r text`无法读到内容，会被阻塞

**Reply by heroxbd** [reply]
> 分享一些我踩过的坑：

特别感谢你的分享！
 
>     3. 使用`$(blablah)`赋值，会将`blablah`的结果中的换行符转成空格

这个观察对了一半，我上课时举了一个例子。

```
heroxbd@proton:/tmp$ s=$(seq 9)
heroxbd@proton:/tmp$ echo $s
1 2 3 4 5 6 7 8 9
heroxbd@proton:/tmp$ echo "$s"
1
2
3
4
5
6
7
8
9
```

**Comment by jiegec** [comment]
WSL 下访问 Windows 权限问题：

如果仓库放在 /mnt/... 目录下，git 会设置为 core.filemode=false 模式，此时创建的文件在 git 中默认是 644 的权限。

解决方法：把仓库移动到 home 目录下，再运行 `git config core.filemode true` 再修改文件权限。
